docker/daemon/client.go

```diff
@@ -0,0 +1,65 @@
+package daemon
+
+import (
+	"github.com/containers/image/types"
+	"github.com/docker/docker/api"
+	dockerclient "github.com/docker/docker/client"
+	"github.com/docker/go-connections/tlsconfig"
+	"net/http"
+	"path/filepath"
+)
+
+// NewDockerClient initializes a new API client based on the passed SystemContext.
+func newDockerClient(systemCtx *types.SystemContext) (*dockerclient.Client, error) {
+	if systemCtx == nil {
+		return dockerclient.NewClient(dockerclient.DefaultDockerHost, api.DefaultVersion, nil, nil)
+	}
+
+	httpClient, err := tlsConfig(systemCtx)
+	if err != nil {
+		return nil, err
+	}
+
+	host := systemCtx.DockerHost
+	if host == "" {
+		host = dockerclient.DefaultDockerHost
+	}
+
+	version := systemCtx.DockerAPIVersion
+	if version == "" {
+		version = api.DefaultVersion
```

Comment:
Why does it make sense to make this overridable? The required API version is determined by the API operations the subpackage calls. Using an older API version would not work, using a newer one should not make a difference.

Cc: @runcom to keep me honest.

Reply:
Using a newer version "should" work, but it might be that in some cases it won't. In these cases setting an explicit version might help. I am trying to mimic what client.NewEnvClient does and there it is possible to set the _DOCKER_API_VERSION_ environment variable to enforce the API version. Since we are not going to use client.NewEnvClient I want to at least provide a solution using  SystemContext which has equivalent functionality.

Reply:
>  but it might be that in some cases it won't. 

Cases like what?

If we don’t _now_ know what this can be good for, I’d rather not add it; it’s extra burden to maintain for no known benefit, and removing it in the future would break API compatibility. We can always add features in the future, with less disruption than removing something, even if useless.

All uses of the environment variable I can find from a quick googling are caused by the client defaulting to a too new version when it doesn’t need those features, and using `DOCKER_API_VERSION` to downgrade to an lower but acceptable version. Using a low enough version in the first place should make `DOCKER_API_VERSION` / this override unnecessary.

Reply:
> If we don’t now know what this can be good for, I’d rather not add it; it’s extra burden to maintain for no known benefit, and removing it in the future would break API compatibility. We can always add features in the future, with less disruption than removing something, even if useless.

I disagree. For me it was important to create something which provides the same functionality as _client.NewEnvClient_. If the programmatic approach cannot offer this, I think we should also add a flag to the context which makes the code use _client.NewEnvClient_.

For what it's worth the Minishift command `minishift docker-env` makes use of the variable as well (I cannot tight now recall the exact details why this got added to the output of this command)

Reply:
Still outstanding.

Reply:
@mtrmac Using a newer one does make a difference, because certain implementations of the Docker engine API simply don't support the old interface at all. Specifically, the Docker engine API on Windows just errors out if you use API version v1.22, making this tool unusable against a Windows Docker host.

Reply:
Whether they actually made enough breaking changes from 1.22 to 1.24 to warrant rejecting any calls made by a client requesting the older version of the API is probably a discussion for a separate thread in https://github.com/moby/moby.

Reply:
That would be a good reason to add some kind of API version detection, based on `docker/docker/client.Client.Ping` (and actually check what _were_ the relevant API changes / version boundaries for our purposes); not a very good reason to give users a “I don’t know what changed and what will break, just do it” hammer _and_ to force users to set an environment variable every time manually.